src/charm.py

```diff
@@ -154,13 +177,87 @@ def _on_ingress_revoked(self, event: IngressPerAppRevokedEvent) -> None:
         if self.unit.is_leader():
             logger.info("This app no longer has ingress")
 
+    def _on_openfga_store_created(self, event: OpenFGAStoreCreateEvent) -> None:
+        """Handle openfga store created event."""
+        if not self._peers:
+            self.unit.status = WaitingStatus("Waiting for peer relation")
+            event.defer()
+            return
+
+        if self._get_openfga_store_info():
+            model_id = self._create_openfga_model()
+            self._set_peer_data(key=self._get_version(), data=dict(openfga_model_id=model_id))
+
+            self._handle_status_update_config(event)
+        else:
+            logger.debug("No openfga store info found, deferring the event")
+            event.defer()
+            return
+
+    def _on_openfga_store_removed(self, event: OpenFGAStoreRemovedEvent) -> None:
+        """Handle openfga store removed event."""
+        logger.info("OpenFGA store was removed")
+        self._pop_peer_data(key=self._get_version())
+        self._handle_status_update_config(event)
+
+    def _on_upgrade_charm(self, event: UpgradeCharmEvent) -> None:
+        """Handle charm upgrade event.
+
+        Create a new model to ensure the migration was run.
+        """
+        if not self._peers:
+            self.unit.status = WaitingStatus("Waiting for peer relation")
+            event.defer()
+            return
+
+        if self._get_openfga_store_info():
+            model_id = self._create_openfga_model()
+            self._set_peer_data(key=self._get_version(), data=dict(openfga_model_id=model_id))
```

Comment:
One thing I am not sure about is whether we should delete the old model. Could we have conflicts between the models? Can openfga handle that?

Reply:
[These](https://openfga.dev/docs/getting-started/immutable-models) docs state that models can't be removed:
> Authorization Models in OpenFGA are immutable, they are created once and then can no longer be deleted or modified.